Integrate the user documentation into the application

return42 opened the pull request:
## What does this PR do?

Integrate the user documentation into the application

Online documentation: rename /help to /doc

## How to test this PR locally?

Build HTML documentation and visit http://0.0.0.0:8000/user/search-syntax.html

    make docs.clean docs.live

Start `make run` session and visit pages:

    http://127.0.0.1:8888/doc/about
    http://127.0.0.1:8888/doc/search-syntax

## Author's checklist

<!-- additional notes for reviewiers -->

## Related issues

- https://github.com/searxng/searxng/pull/872
- #722 
- #659

Closes: #659, #722

not-my-profile committed:
[docs] move own-instance.rst out of docs/user

not-my-profile committed:
[help] move about.md to searx/help/en/

dalf commented on manage:
```diff
@@ -419,6 +419,7 @@ docs.prebuild() {
         ./utils/searx.sh doc   | cat > "${DOCS_BUILD}/includes/searx.rst"
         ./utils/filtron.sh doc | cat > "${DOCS_BUILD}/includes/filtron.rst"
         ./utils/morty.sh doc   | cat > "${DOCS_BUILD}/includes/morty.rst"
+        pyenv.cmd python -m searx.user_help search-syntax > docs/user/search-syntax.md
```
If we need to change the `manage` each time there is a new page, that's not scalable.
In `searx.user_help`, why not parse everything in the TOC?

return42 replied:
> why not parse everything in the TOC

I have a different intention about sharing pages from online doc in project doc / see https://github.com/searxng/searxng/pull/930#discussion_r820058125

dalf commented on searx/help/en/search-syntax.md:
```diff
@@ -0,0 +1,35 @@
+search in map category **and** duckduckgo **and** wikipedia for
+`paris`.
+
+See the {{link('preferences', 'preferences')}} for the list of engines,
```
It requires translators to deal Markdown and Jinja2: they have to know that the first `preferences` can be translated, not the second one.

`[preferences]({{ link('preferences') }})` is complicated

return42 replied:
That is true / I did not had an idea how to make it smarter for translators.

For search links we have:

    {{search('!wp paris')}}

Instead of:

    {{link('preferences', 'preferences')}}

we can use:

    {{link('preferences --> preferences')}}

but I do not know if it is better / any other suggestion for a markup of links that needed to be rendered in a Flask context?

dalf commented on searx/user_help.py:
```diff
@@ -24,7 +25,7 @@ class HelpPage(NamedTuple):
 
 # Whenever a new .md file is added to help/ it needs to be added here
-_TOC = ('about',)
+_TOC = ('search-syntax', 'about')
```
To put the about page as the first tab:
`_TOC = ('about', 'search-syntax')`

return42 replied:
IMO users will more often look for the "search syntax" then for "about", that was my reason to place it this order, but I don't really care.  We can use the order you prefer.

dalf commented on searx/webapp.py:
```diff
@@ -896,19 +897,28 @@ def __get_translated_errors(unresponsive_engines: Iterable[UnresponsiveEngine]):
-def help_page(pagename):
-    """Render help page"""
-    page = user_help.PAGES.get(pagename)
+@app.route('/doc/<pagename>', methods=['GET'])
```
IMO it is important to keep the language in URL, otherwise different contents are going to be shown for the same URL: this is going to lead to confusion when the links are shared.

return42 replied:
Can we improve the routes in another PR? .. I want to touch to much in this PR and this PR already touches a lot.

dalf commented on searx/doc.py:
```diff
@@ -0,0 +1,290 @@
+    # found ..
+    import time
+
+    time.sleep(3)
```
If I understand correctly, the typical use case is:
* we load the app to resolve the URL like `/search`.
* we render the page by discarding the URL (we don't need the app).
* once rendered, we wait for the app to finish the background initialization

So do we load the application?

return42 replied:
> So do we load the application?

You asking "why do we load the app"? ..  because the `url_for` command needs the  flask context. See the comment above https://github.com/searxng/searxng/pull/930/files#diff-b439f9f6bc445fcbb1be1c7d0314a845126d56ba235e52361c9918d943d5733aR84-R90


```
    # To render the online documentation a Flask application context is needed.
    # The url_for functions in the jinja templates need all routes to be
    # registered in the Flask app.
```

dalf replied:
Call sequence:
* `pyenv.cmd python -m searx.doc search-syntax Offline`
* Pick the [Offline](https://github.com/searxng/searxng/blob/5d45cd6eb18095e0bb3e5715c58072e68c4aa727/searx/doc.py#L64)
* [`main()`: `doc = OnlineDoc(Offline())`](https://github.com/searxng/searxng/blob/5d45cd6eb18095e0bb3e5715c58072e68c4aa727/searx/doc.py#L75)
* [`main()`:  Initialize `searx.webapp`](https://github.com/searxng/searxng/blob/5d45cd6eb18095e0bb3e5715c58072e68c4aa727/searx/doc.py#L93-L96)
* [`main()`: `page = doc.render(pagename)`](https://github.com/searxng/searxng/blob/5d45cd6eb18095e0bb3e5715c58072e68c4aa727/searx/doc.py#L102-L104)
* [`OnlineDoc.render(...)`: `page = self.markup.render(content)`](https://github.com/searxng/searxng/blob/5d45cd6eb18095e0bb3e5715c58072e68c4aa727/searx/doc.py#L283)
* [`Markup.render(...)`: `ctx = self.get_ctx()`](https://github.com/searxng/searxng/blob/5d45cd6eb18095e0bb3e5715c58072e68c4aa727/searx/doc.py#L153-L157)
* [`Offline.get_ctx()`](https://github.com/searxng/searxng/blob/5d45cd6eb18095e0bb3e5715c58072e68c4aa727/searx/doc.py#L228-L235) : here we don't use `url_for`
* [`Markup.render(...)`: `return template.render(**ctx)`](https://github.com/searxng/searxng/blob/5d45cd6eb18095e0bb3e5715c58072e68c4aa727/searx/doc.py#L157) : so `url_for` won't be called, so there is no need for the app to be initialized.

return42 replied:
> so `url_for` won't be called, so there is no need for the app to be initialized.

Not for the `url_for` (`link` & `search`) function but may be in the future we want to use other information or functions from the Flask context? .. if we will never use any information from the Flask context in the project documentation (aka Offline) .. we can simplify the code a lot (like you have done in your branch).

What do you think, should I simplify the code of this PR?

dalf replied:
> should I simplify the code

Yes!

Currently, I feel `searx/doc.py` is a tool that should be in `searxng_extra`.
IMO, for the runtime, there are too many abstractions.

return42 replied:
Hm .. just I have another thought (sorry) .. I think the Flask context is also useful when the project documentation is build.  Assume the `base_url` is set:

https://github.com/searxng/searxng/blob/927aa71133839ce5713b742b43285953301adf36/searx/settings.yml#L45

then we can also compile links in the the project documentation / that is linked by

https://github.com/searxng/searxng/blob/927aa71133839ce5713b742b43285953301adf36/searx/settings.yml#L9

Since we do not have a base_url in our project, we do not need this, but forks who build their own online documentation will find it useful I think.

With said in mind, what do you think about?

----

EDIT:

>  Currently, I feel `searx/doc.py` is a tool that should be in `searxng_extra`.
> IMO, for the runtime, there are too many abstractions.

In a rework of this PR I can simplify the code and see what we can move to `searxng_extra` .. but first I want to clarify with you if it make sense to have a flask context in the build process of the project doc.

dalf replied:
I'm not sure to understand.

Do you suggest that some forks may wish to create links from the Sphinx documentation to their own instance?
So in this case `url_for` would be useful for them.

[EDIT] I can't say, but I have NOT seen ANY requests in that way and so let's skip this for now.

return42 replied:
> Do you suggest that some forks may wish to create links from the Sphinx documentation to their own instance?

yes.

dalf commented on searx/doc.py:
```diff
@@ -0,0 +1,290 @@
+        self.l10n: typing.List = [
+            'en',
+        ]
+        self.toc: typing.List = ['search-syntax', 'about']
```
Related to the change in `./manage` : the TOC (*) is declared twice in two different locations.

(*) once including `about`, once without.

return42 replied:
> (*) once including `about`, once without.

My intention is different to yours: IMO both documentations are not related to each other! There will be only some files in the online documentation we want to share in the project documentation.

IMO 'search-syntax' and 'about' is what we have now but I think in the future we will rework the about page and we will have more pages and some of them will not be a part of the project documentation.

This PR is just the starting point to share files from the online doc in the project doc.

dalf replied:
Can we avoid multiple declaration of the same information?

Currently, there are lists in:
* `searx/doc.py`
* `manage`
* `docs/user/.gitignore`

The last two are the same, the first one is different according to what you describe.

My idea is to include all the pages and includes two sections in `index.rst`: 
   * one with generic content (syntax.md, ...)
   * one with the documentation dedicated for the instances (about.md, etc...). Alternative: we just ignore these files (there are compiled but not included in the TOC).
--> Most importantly: in one .rst file, we can decide what belongs to the Sphinx documentation.

---

This is related to https://github.com/searxng/searxng/issues/808

return42 replied:
> Most importantly: in one .rst file, we can decide what belongs to the Sphinx documentation.

If you place .md files in the /doc but don't include them in the TOC, the Sphinx-doc parser will warn you that there are files not included in any TOC .. may be this warnings can be suppressed by a conf.py setting .. but then we did have won anything about your initial question:

> Can we avoid multiple declaration of the same information?

dalf reviewed (commented):
See the comments.
From them, I've made the commit https://github.com/searxng/searxng/commit/e5bc94f07afff465580acc9c15fb505585369881 in this branch: https://github.com/dalf/searxng/tree/user-doc-3

dalf commented on searx/webapp.py:
```diff
@@ -896,19 +897,28 @@ def __get_translated_errors(unresponsive_engines: Iterable[UnresponsiveEngine]):
     return render(
-        'help.html', page=user_help.PAGES[pagename], all_pages=user_help.PAGES.items(), page_filename=pagename
+        'doc.html',
+        all_pages=_ONLINE_DOC.all_pages,
```
The call to `all_pages` is going to render all the pages for the selected language.

Why not make the code simpler:
* pre-render all the pages when the application starts.
* server the rendered page from this endpoint

return42 replied:
> The call to `all_pages` is going to render all the pages for the selected language.

Yes, we need the titles (in the tabs) of all pages in the selected languages.

> pre-render all the pages when the application starts.

IMO rendering Markdown to HTML is an *expensive* task .. the parser needs to be started and the output (of all pages in all languages) need to be rendered.  I think a lazy rendering on demand combined with caching is much *cheaper* since users will not visit very often the online documentation and IMO no user will read all online doc pages in all languages.

dalf replied:
> IMO rendering Markdown to HTML is an expensive task ..

To have an idea, https://github.com/jlevy/the-art-of-command-line/blob/master/README.md (40KB) takes few seconds on my laptop (slow mode, power cable unplugged). Multiple by the number of the language this can be a problem if we have the same amount of document (once again, we don't know what will be the user documentation which makes the decision taking not easy).

The rendering time issue exists also in this PR, because any request to the documentation is going to render all pages  for this language.

Call sequence:
* https://github.com/searxng/searxng/blob/5d45cd6eb18095e0bb3e5715c58072e68c4aa727/searx/webapp.py#L907
* https://github.com/searxng/searxng/blob/5d45cd6eb18095e0bb3e5715c58072e68c4aa727/searx/doc.py#L266
* https://github.com/searxng/searxng/blob/5d45cd6eb18095e0bb3e5715c58072e68c4aa727/searx/webapp.py#L917-L922
* https://github.com/searxng/searxng/blob/5d45cd6eb18095e0bb3e5715c58072e68c4aa727/searx/templates/simple/doc.html#L5
* https://github.com/searxng/searxng/blob/5d45cd6eb18095e0bb3e5715c58072e68c4aa727/searx/doc.py#L260
* for each pages of a language: 
  * https://github.com/searxng/searxng/blob/5d45cd6eb18095e0bb3e5715c58072e68c4aa727/searx/doc.py#L263
  * https://github.com/searxng/searxng/blob/5d45cd6eb18095e0bb3e5715c58072e68c4aa727/searx/doc.py#L266 

Idea:
* load the raw content of all the pages
* parse the raw content to get the title and the markdown content
* then do lazy rendering of the markdown content

So we have real lazy rendering.

return42 replied:
> So we have real lazy rendering.

Sound good to me, I will give it a try in a rewrite of this PR.

dalf commented on docs/user/.gitignore:
```diff
@@ -0,0 +1 @@
+search-syntax.md
```
Would `*.md` be better?
So when there is a new page, there is no need to update this file

return42 replied:
See my comment above https://github.com/searxng/searxng/pull/930#discussion_r820058125

dalf commented on searx/doc.py:
```diff
@@ -0,0 +1,290 @@
+        return ctx
+
+
+_DOC_FOLDER = os.path.join(os.path.abspath(os.path.dirname(__file__)), 'doc')
```
Is there a problem to use the `pkg_resources` instead of the `os` package and `open`?

return42 replied:
My intention is different.  IMO (not in this PR but may be in the future) the location is taken from settings.yml / see https://github.com/searxng/searxng/pull/930/files#diff-b439f9f6bc445fcbb1be1c7d0314a845126d56ba235e52361c9918d943d5733aR252-R258

```
        # future: could be set from settings.xml
        self.folder: str = _DOC_FOLDER
        self.i18n_origin: str = 'en'
        self.l10n: typing.List = [
            'en',
        ]
        self.toc: typing.List = ['search-syntax', 'about']
```

By this instance maintainer can write their own online documentation.  By example in my instance I use *hostname replace* to link all youtube results to invidious instances .. and I want to document this setup to my users.

dalf replied:
>  I use hostname replace to link all youtube results to invidious instances

This need exists for everyone using this plugin.
IMO, the plugin should provide a documentation or at least some information in the `/preferences` page (there is no way to do the later currently)

return42 replied:
This plugin was just an example .. I think there is a need for having a online documentation to describe the intention of the setup of that instance.

By example: [... but all results (incl. yahoo) are redirected to an instance in DE where the majority of my users is located](https://github.com/searxng/searxng/discussions/911)

dalf replied:
> By example: https://github.com/searxng/searxng/discussions/911

Where this documentation should be?
Currently it requires to change `searx/doc.py` to include the additional pages in the tab.

Idea:
* keep the git repository clean
* allow to add additional documentation repository: all pages declared here are included in the tab.

So the administrator just have to declare the directory, add some files nothing more.

(perhaps it could be better, simpler)

return42 replied:
> So the administrator just have to declare the directory, add some files nothing more.

Sounds good to me .. 

In this PR I want to prepare a boilerplate that is able to do so, but I would like to implement the functionality (where the values are taken from settings.yml) in another PR .. is it OK for you?

dalf replied:
IMO, let's do that in another PR and keep the focus on the rendering of the `searx/doc` content.

return42 commented on requirements-dev.txt:
```diff
@@ -8,12 +8,14 @@ splinter==0.17.0
 sphinx-issues==3.0.1
-sphinx-jinja==1.4.0
-sphinx-tabs==3.2.0
+sphinx-jinja @ git+http://github.com/return42/sphinx-jinja.git
```
Interim solution / see https://github.com/tardyp/sphinx-jinja/pull/30 and https://github.com/tardyp/sphinx-jinja/issues/28#issuecomment-1060701874

return42 commented on requirements-dev.txt:
```diff
@@ -8,12 +8,14 @@ splinter==0.17.0
-sphinx-jinja==1.4.0
-sphinx-tabs==3.2.0
+sphinx-jinja @ git+http://github.com/return42/sphinx-jinja.git
+sphinx-tabs @ git+https://github.com/executablebooks/sphinx-tabs
```
Interim solution / see https://github.com/executablebooks/sphinx-tabs/pull/144#issuecomment-1060717377

dalf commented on searx/doc.py:
```diff
@@ -0,0 +1,199 @@
+        if self._raw_content is None:
+            with open(self.fname, 'r', encoding='utf-8') as f:
+                self._raw_content = f.read()
+        return self._raw_content
```
With [cached_property](https://docs.python.org/3/library/functools.html#functools.cached_property) :
```python
from functools import cached_property

...
    @cached_property
    def raw_content(self):
        """Raw content of the page (without any jinja rendering)"""
        with open(self.fname, 'r', encoding='utf-8') as f:
            return f.read()
```

dalf commented on searx/doc.py:
```diff
@@ -0,0 +1,199 @@
+        self.i18n_origin: str = 'en'
+        """default language"""
+
+        self.l10n: typing.List = [
```
AFIU, `l10n` means `localisation`, here the values are languages.

return42 replied:
`localisation` refers to the adaptation of software and documentation to a specific country .. we use the term "locale" very often when we mean a language.  But if you want, I can rename it to `languages`.

dalf replied:
in settings.yml and preferences.py :
* `locale` is the language UI + region most probably [because of babel](https://github.com/searxng/searxng/blob/e1319fc21c713cfb9675389ba38f37e10cf77cd2/searx/webapp.py#L257)
* we will have to `zh_Hans_CN` and `zh_Hant_TW` for example.
* `languages` is the query language (in preferences.py and webapp.py)

So I vote for `self.locale`

dalf commented on searx/doc.py:
```diff
@@ -0,0 +1,199 @@
+        self.folder: str = _DOC_FOLDER
+        """location of the Markdwon files"""
+
+        self.i18n_origin: str = 'en'
```
Why not `default_language` or `fallback_language`?

return42 replied:
Because this is the origin document, the other languages are just translations.

dalf replied:
weblate uses ["source language"](https://docs.weblate.org/en/weblate-4.11/admin/projects.html#component-source-language).

Since we use locale everywhere (except for the query language), I suggest `self.locale_source`

dalf commented on searxng_extra/docs_prebuild:
```diff
@@ -0,0 +1,83 @@
+    return DOC
+
+
+def _render_all_with_flask_ctx(base_url):
```
Untested simpler version:

```python
#!/usr/bin/env python
# lint: pylint
# SPDX-License-Identifier: AGPL-3.0-or-later

"""Script that implements some prebuild tasks needed by target docs.prebuild
"""

import sys
import os.path
import time
import typing

from searx import settings, get_setting, doc

_doc_user = os.path.abspath(os.path.join(os.path.dirname(__file__), '..', 'docs', 'user'))


class OfflinePage(doc.DocPage):

    def get_ctx(self):  # pylint: disable=no-self-use
        """Jinja context to render :py:obj:`DocPage.content` for offline purpose (no
        links to SearXNG instance)"""

        ctx = super().get_ctx()
        ctx['link'] = lambda name, url: '`%s`' % name
        ctx['search'] = lambda query: '`%s`' % query

        return ctx


class OfflineDoc(doc.OnlineDoc):

    def __init__(self, base_url):
        super().__init__(doc.DocPage)
        self.base_url = base_url  # so the base_url parameter can be removed from OnlineDoc

    def all_pages(self, locale: typing.Optional[str] = None):
        settings['server']['secret_key'] = "x"
        from searx.webapp import app

        with app.test_request_context(base_url=self.base_url):
            yield from super().all_pages(locale)

        time.sleep(3)


def main():
    base_url = get_setting('server.base_url', None)
    DOC = OfflineDoc(base_url) if base_url else doc.OnlineDoc(OfflinePage)
    for pagename, page in DOC.all_pages('en'):
        fname = os.path.join(_doc_user, os.path.basename(page.fname))
        with open(fname, 'w') as f:
            f.write(page.content)


if __name__ == '__main__':
    sys.exit(main())
```

dalf commented on searx/doc.py:
```diff
@@ -0,0 +1,199 @@
+_DOC_FOLDER = os.path.join(os.path.abspath(os.path.dirname(__file__)), 'doc')
+
+
+class OnlineDoc:  # pylint: disable=too-few-public-methods
```
With the line in mind ( from `searxng_extra/docs_prebuild` ):
```
DOC = doc.OnlineDoc(OfflinePage)  # It requires a careful reading to understand that there is no cat in a box that may be alive or dead.
```

`UserDoc`, `UserDocumentation` or a name describing the content would be better.

return42 committed:
[mod] replace /help by /info pages and include pages in project docs

This patch implements a bolierplate to share content from info-pages of the
SearXNG instance (URL /info) with the project documentation (path /docs/user).

The info pages are using Markdown (CommonMark), to include them in the project
documentation (reST) the myst-parser [1] is used in the Sphinx-doc build chain.

If base_url is known (defined in settings.yml) links to the instance are also
inserted into the project documentation::

    searxng_extra/docs_prebuild

[1] https://www.sphinx-doc.org/en/master/usage/markdown.html

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[simple] add CSS for /info pages

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[interim fix] of sphinx-tabs and sphinx-jinja

The myst-parser requires >= docutils v.0.17 what ends in a dependency hell where
plugins sphinx-tabs and sphinx-jinja we use are involved.

This patch can be reverted when [2], [3], [4]  are solved and new release is
available / see [1].

[1] https://github.com/searxng/searxng/pull/954
[2] https://github.com/executablebooks/sphinx-tabs/issues/152
[3] https://github.com/executablebooks/sphinx-tabs/pull/153
[4] https://github.com/executablebooks/sphinx-tabs/pull/154

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[build] /static

return42 committed:
[fix] add module for backward compatibility

cache_property has been added in py3.8 [1]

To support cache_property in py3.7 the implementation from 3.8 has been
copied to compat.py.  This code can be cleanup with EOL of py3.7.

[1] https://docs.python.org/3/library/functools.html#functools.cached_property

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

dalf committed:
Various change on PR 930

dalf commented on searx/static/themes/simple/src/less/info.less:
```diff
@@ -0,0 +1,13 @@
+.info-page {
+  font-family: sans-serif;
+  font-size: 1.3em;
```
`1em` seems better IMO

dalf commented on searx/templates/simple/base.html:
```diff
@@ -58,7 +58,7 @@
   <footer>
     <p>
-    {{ _('Powered by') }} <a href="{{ url_for('help_page', pagename='about') }}">searxng</a> - {{ searx_version }} — {{ _('a privacy-respecting, hackable metasearch engine') }}<br/>
+    {{ _('Powered by') }} <a href="{{ url_for('info', pagename='about') }}">searxng</a> - {{ searx_version }} — {{ _('a privacy-respecting, hackable metasearch engine') }}<br/>
```
There is no `locale` here : `url_for_theme` is going to add the locale, so even in a 404 page, the link is going to work.

dalf commented on searx/webapp.py:
```diff
@@ -382,6 +383,11 @@ def url_for_theme(endpoint: str, override_theme: Optional[str] = None, **values)
         if file_hash:
             values['filename'] = filename_with_theme
             suffix = "?" + file_hash
+    if endpoint == 'info' and 'locale' not in values:
```
see `simple/base.html`  and `oscar/nav.html`.

allow to have a working link to the about page even in a 404 page.

dalf commented on searx/webapp.py:
```diff
@@ -898,19 +906,36 @@ def __get_translated_errors(unresponsive_engines: Iterable[UnresponsiveEngine]):
     if page is None:
         flask.abort(404)
 
+    def all_pages():
```
Iterator over all the page of the TOC including:
* pagename
* page
* locale

If the locale is `de` and `de/about.md` exists (and no other translated page), this function is going to yield:
* `search_syntax`, `PageInfo(....)`, `en`
* `about`, `PageInfo(....)`, `de`

dalf commented on searxng_extra/docs_prebuild:
```diff
@@ -0,0 +1,84 @@
+        for _, page in infopageset.all_pages('en'):
+            fname = os.path.join(_doc_user, os.path.basename(page.fname))
+            with open(fname, 'w') as f:
+                f.write(page.content)
```
no need to call `render()`, we just need the markdown content.

dalf commented on searx/infopage/__init__.py:
```diff
@@ -0,0 +1,179 @@
+
+        fname = os.path.join(self.folder, locale, pagename) + '.md'
+        if not os.path.exists(fname):
+            logger.info('file %s does not exists', fname)
```
A translator doesn't have to translate all the pages: this is not an error if a page doesn't exist. The `info` endpoint in `webapp.py` deals with missing pages.

dalf commented on searx/infopage/__init__.py:
```diff
@@ -0,0 +1,179 @@
+        fname = os.path.join(self.folder, locale, pagename) + '.md'
+        if not os.path.exists(fname):
+            logger.info('file %s does not exists', fname)
+            self.CACHE[cache_key] = None
```
Cache the fact the page was not found

dalf commented on searx/infopage/__init__.py:
```diff
@@ -0,0 +1,179 @@
+        return t
+
+    @cached_property
+    def html(self):
```
Remove the render function and keep only a `html` property.

dalf commented on searx/infopage/__init__.py:
```diff
@@ -0,0 +1,179 @@
+
+        return ctx
+
+    def __repr__(self):
```
Might be useful for debugging.

dalf commented on searx/infopage/__init__.py:
```diff
@@ -0,0 +1,179 @@
+        """Jinja context to render :py:obj:`InfoPage.content`"""
+
+        def _md_link(name, url):
+            url = url_for(url, _external=True)
```
Use `_external`, so there is no need to add a `base_url` as parameter of the constructor of the class.

dalf commented on searx/infopage/__init__.py:
```diff
@@ -0,0 +1,179 @@
+        self.locale_default: str = 'en'
+        """default language"""
+
+        self.locales: typing.List = [locale for locale in os.listdir(_INFO_FOLDER) if locale in LOCALE_NAMES]
```
Read the existing directories and make sure there are actual locales: a new translation doesn't require to change the code.

dalf commented on searx/infopage/__init__.py:
```diff
@@ -0,0 +1,179 @@
+
+        # future: could be set from settings.xml
+
+        self.folder: str = info_folder or _INFO_FOLDER
```
Tests don't need to hook _INFO_FOLDER (tests... that are not written)

dalf commented on searx/infopage/__init__.py:
```diff
@@ -0,0 +1,179 @@
+        self.CACHE[cache_key] = page
+        return page
+
+    def all_pages(self, locale: typing.Optional[str] = None):
```
Simple iterator here: the complexity is moved to `searx.webapp` because it requires various parameter from the current HTTP request. May be not the best solution.

dalf commented on searx/infopage/__init__.py:
```diff
@@ -0,0 +1,179 @@
+        :type locale: str
+
+        """
+        locale = locale or self.locale_default
```
Fallback to the default value (before the fallback was set after `if locale not in self.locales` which is always `False` if `locale is None`

dalf commented on searx/infopage/__init__.py:
```diff
@@ -0,0 +1,179 @@
+_INFO_FOLDER = os.path.abspath(os.path.dirname(__file__))
+
+
+class InfoPage:
```
Keep only one class for the pages: `render()` was never called in `InfoPage`.

dalf commented on docs/user/index.rst:
```diff
@@ -2,9 +2,14 @@
 User documentation
```
We should not name the same thing with different names.
`User documentation` must be replaced by `Information` or `User Information` if `Information` is the name in the webapp.

dalf reviewed (commented):
I've added a commit and some comments to explain the changes.
